File: src/FullBuild/Actions/Workspace.cs
// Copyright (c) 2014, Pierre Chalamet
// All rights reserved.
// 
// Redistribution and use in source and binary forms, with or without
// modification, are permitted provided that the following conditions are met:
//     * Redistributions of source code must retain the above copyright
//       notice, this list of conditions and the following disclaimer.
//     * Redistributions in binary form must reproduce the above copyright
//       notice, this list of conditions and the following disclaimer in the
//       documentation and/or other materials provided with the distribution.
//     * Neither the name of Pierre Chalamet nor the
//       names of its contributors may be used to endorse or promote products
//       derived from this software without specific prior written permission.
// 
// THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS" AND
// ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED
// WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
// DISCLAIMED. IN NO EVENT SHALL PIERRE CHALAMET BE LIABLE FOR ANY
// DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES
// (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
// LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND
// ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
// (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
// SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.

using System;
using System.Collections.Generic;
using System.Collections.Immutable;
using System.IO;
using System.Linq;
using System.Reflection;
using System.Xml.Linq;
using FullBuild.Config;
using FullBuild.Helpers;
using FullBuild.Model;
using FullBuild.SourceControl;
using Newtonsoft.Json;
using Newtonsoft.Json.Linq;
using NLog;

namespace FullBuild.Actions
{
    internal class Workspace
    {
        private static readonly Logger _logger = LogManager.GetCurrentClassLogger();

        public void Update()
        {
            var workspace = WellKnownFolders.GetWorkspaceDirectory();
            var config = ConfigManager.GetConfig(workspace);

            // get all csproj in all repos only
            var anthology = LoadOrCreateAnthology();
            anthology = UpdateAnthologyFromSource(config, workspace, anthology);
            Dump(anthology);

            // Promotion
            anthology = PromoteToProject(anthology);

            // merge with existing
            anthology = MergeNewAnthologyWithExisting(anthology);

            // Generate import files
            GenerateImports(anthology);
        }

        private Anthology PromoteToProject(Anthology anthology)
        {
            var pkgsDir = WellKnownFolders.GetPackageDirectory();

            // 1. remove binary reference imported from packages
            foreach(var project in anthology.Projects)
            {
                // gather all assemblies from packages in this project
                var importedAssemblies = (from pkgRef in project.PackageReferences
                                          let pkgdir = pkgsDir.GetDirectory(pkgRef)
                                          where pkgdir.Exists
                                          select Nuspec.Assemblies(pkgdir)).SelectMany(x => x).Distinct(StringComparer.InvariantCultureIgnoreCase);

                // remove imported assemblies
                var newProject = importedAssemblies.Aggregate(project, (p, a) => p.RemoveBinaryReference(a));
                anthology = anthology.AddOrUpdateProject(newProject);
            }

            // 2. promote package to project
            var pkg2prj = from package in anthology.Packages
                          let pkgdir = pkgsDir.GetDirectory(package.Name)
                          where pkgdir.Exists
                          let assemblies = Nuspec.Assemblies(pkgdir)
                          from project in anthology.Projects
                          where assemblies.Contains(project.AssemblyName, StringComparer.InvariantCultureIgnoreCase)
                          select new {Pkg = package, Prj = project};

            foreach(var p2p in pkg2prj)
            {
                _logger.Info("Migrating package {0} to project {1}", p2p.Pkg.Name, p2p.Prj.ProjectFile);
                foreach(var project in anthology.Projects)
                {
                    if (project.PackageReferences.Contains(p2p.Pkg.Name, StringComparer.InvariantCultureIgnoreCase))
                    {
                        var newProject = project.RemovePackageReference(p2p.Pkg.Name);
                        newProject = newProject.AddProjectReference(p2p.Prj.Guid);
                        anthology = anthology.AddOrUpdateProject(newProject);
                    }
                }
            }

            // 3. remove empty package
            var emptyPackages = from package in anthology.Packages
                                let pkgdir = pkgsDir.GetDirectory(package.Name)
                                where pkgdir.Exists
                                let assemblies = Nuspec.Assemblies(pkgdir)
                                where !assemblies.Any()
                                select package;

            foreach(var project in anthology.Projects)
            {
                var newProject = emptyPackages.Aggregate(project, (p, pa) => p.RemovePackageReference(pa.Name));
                anthology = anthology.AddOrUpdateProject(newProject);
            }

            anthology = emptyPackages.Aggregate(anthology, (a, p) => a.RemovePackage(p));

            // 4. remove unused package
            var usedPackages = anthology.Projects.SelectMany(x => x.PackageReferences).Distinct();

            var packagesToRemove = from package in anthology.Packages
                                   where !usedPackages.Contains(package.Name, StringComparer.InvariantCultureIgnoreCase)
                                   select package;

            anthology = packagesToRemove.Aggregate(anthology, (a, p) => a.RemovePackage(p));

            // 5. remove unused binary
            var usedBinaries = anthology.Projects.SelectMany(x => x.BinaryReferences).Distinct();

            var binariesToRemove = from binary in anthology.Binaries
                                   where !usedBinaries.Contains(binary.AssemblyName, StringComparer.InvariantCultureIgnoreCase)
                                   select binary;

            anthology = binariesToRemove.Aggregate(anthology, (a, b) => a.RemoveBinary(b));

            return anthology;
        }

        public void Init(string path)
        {
            var wsDir = new DirectoryInfo(path);
            wsDir.Create();

            var admDir = wsDir.GetDirectory(".full-build");
            if (admDir.Exists)
            {
                throw new ArgumentException("Workspace is already initialized");
            }

            var config = ConfigManager.GetConfig(wsDir);

            var sourceControl = ServiceActivator<Factory>.Create<ISourceControl>(config.SourceControl);
            sourceControl.Clone("full-build", config.AdminRepo, admDir);

            // force a .gitignore
            //var gitIgnoreFile = admDir.GetFile(".gitignore");
            //var gitIgnore = new StringBuilder().AppendLine("cache").AppendLine("projects").AppendLine("views").ToString();
            //File.WriteAllText(gitIgnoreFile.FullName, gitIgnore);

            config = ConfigManager.GetConfig(wsDir);
            foreach(var repo in config.SourceRepos)
            {
                var repoDir = wsDir.GetDirectory(repo.Name);
                sourceControl.Clone(repo.Name, repo.Url, repoDir);
            }
        }

        private void GenerateImports(Anthology anthology)
        {
            var targetDir = WellKnownFolders.GetProjectDirectory();
            targetDir.Create();

            foreach(var project in anthology.Projects)
            {
                var projectFile = Path.Combine(WellKnownFolders.MsBuildSolutionDir, project.ProjectFile);
                var binFile = Path.Combine(WellKnownFolders.MsBuildBinDir, project.AssemblyName + project.Extension);
                var srcCondition = string.Format("'$({0}_Src)' != ''", project.AssemblyName.Replace('.', '_'));
                var binCondition = string.Format("'$({0}_Src)' == ''", project.AssemblyName.Replace('.', '_'));

                var xdoc = new XElement(XmlHelpers.NsMsBuild + "Project",
                                        new XElement(XmlHelpers.NsMsBuild + "Import",
                                                     new XAttribute("Project", Path.Combine(WellKnownFolders.MsBuildViewDir, "$(SolutionName).targets")),
                                                     new XAttribute("Condition", "'$(BinSrcConfig)' == ''")),
                                        new XElement(XmlHelpers.NsMsBuild + "ItemGroup",
                                                     new XElement(XmlHelpers.NsMsBuild + "ProjectReference",
                                                                  new XAttribute("Include", projectFile),
                                                                  new XAttribute("Condition", srcCondition),
                                                                  new XElement(XmlHelpers.NsMsBuild + "Project", project.Guid.ToString("B")),
                                                                  new XElement(XmlHelpers.NsMsBuild + "Name", project.AssemblyName)),
                                                     new XElement(XmlHelpers.NsMsBuild + "Reference",
                                                                  new XAttribute("Include", project.AssemblyName),
                                                                  new XAttribute("Condition", binCondition),
                                                                  new XElement(XmlHelpers.NsMsBuild + "HintPath", binFile))));

                var targetFileName = project.Guid + ".targets";
                var prjImport = targetDir.GetFile(targetFileName);
                xdoc.Save(prjImport.FullName);
            }
        }

        private static Anthology UpdateAnthologyFromSource(FullBuildConfig config, DirectoryInfo workspace, Anthology anthology)
        {
            _logger.Info("Anthology contains {0} projects", anthology.Projects.Count());
            foreach(var repo in config.SourceRepos)
            {
                _logger.Info("Processing repo {0}", repo.Name);
                var repoDir = workspace.GetDirectory(repo.Name);

                // delete all solution files
                var slns = repoDir.EnumerateFiles("*.sln", SearchOption.AllDirectories);
                slns.ForEach(x => x.Delete());

                // process all projects
                var csprojs = repoDir.EnumerateFiles("*.csproj", SearchOption.AllDirectories);
                _logger.Info("Found {0} projects", csprojs.Count());

                anthology = csprojs.Aggregate(anthology, (a, p) => ParseAndAddProject(workspace, p, a));

                _logger.Info("Project count {0}", anthology.Projects.Count());
            }

            return anthology;
        }

        private static Anthology LoadOrCreateAnthology()
        {
            var admDir = WellKnownFolders.GetAdminDirectory();
            var anthologyFile = admDir.GetFile(Anthology.AnthologyFileName);
            if (anthologyFile.Exists)
            {
                var oldJson = File.ReadAllText(anthologyFile.FullName);
                var prevAnthology = JsonConvert.DeserializeObject<Anthology>(oldJson);
                var oldJ = JObject.FromObject(prevAnthology);
                var anthology = oldJ.ToObject<Anthology>();
                return anthology;
            }

            return new Anthology();
        }

        private static Anthology MergeNewAnthologyWithExisting(Anthology anthology)
        {
            // merge anthology files
            var admDir = WellKnownFolders.GetAdminDirectory();
            var anthologyFile = admDir.GetFile(Anthology.AnthologyFileName);
            if (anthologyFile.Exists)
            {
                var oldJson = File.ReadAllText(anthologyFile.FullName);
                var prevAnthology = JsonConvert.DeserializeObject<Anthology>(oldJson);
                var newJ = JObject.FromObject(anthology);
                var oldJ = JObject.FromObject(prevAnthology);

                var mergeSettings = new JsonMergeSettings {MergeArrayHandling = MergeArrayHandling.Replace};
                oldJ.Merge(newJ, mergeSettings);
                anthology = oldJ.ToObject<Anthology>();
            }

            var json = JsonConvert.SerializeObject(anthology, Formatting.Indented);
            File.WriteAllText(anthologyFile.FullName, json);

            return anthology;
        }

        private static IEnumerable<Model.Package> GetNugetPackages(DirectoryInfo projectDir)
        {
            var packageFile = projectDir.GetFile("packages.config");
            if (!packageFile.Exists)
            {
                return Enumerable.Empty<Model.Package>();
            }

            var docPackage = XDocument.Load(packageFile.FullName);
            var packages = from element in docPackage.Descendants("package")
                           let name = (string) element.Attribute("id")
                           let version = (string) element.Attribute("version")
                           select new Model.Package(name, version);

            return packages;
        }

        private static void Dump(Anthology anthology)
        {
            foreach(var project in anthology.Projects)
            {
                _logger.Debug("Consistency for project {0}", project.GetName());

                foreach(var projectDependency in project.ProjectReferences)
                {
                    var target = anthology.Projects.Single(x => x.Guid == projectDependency);
                    _logger.Debug(" --> {0}", target.GetName());
                }

                foreach(var binaryDependency in project.BinaryReferences)
                {
                    _logger.Debug(" ++ {0}", binaryDependency);
                }
            }
        }

        private static Anthology ParseAndAddProject(DirectoryInfo workspace, FileInfo projectFile, Anthology anthology)
        {
            var xdoc = XDocument.Load(projectFile.FullName);

            var projectFileName = projectFile.FullName.Substring(workspace.FullName.Length + 1);

            var projectGuid = Guid.ParseExact((string) xdoc.Descendants(XmlHelpers.NsMsBuild + "ProjectGuid").Single(), "B");
            var assemblyName = (string) xdoc.Descendants(XmlHelpers.NsMsBuild + "AssemblyName").Single();
            var fxTarget = (string) xdoc.Descendants(XmlHelpers.NsMsBuild + "TargetFrameworkVersion").Single();

            var extension = ((string) xdoc.Descendants(XmlHelpers.NsMsBuild + "OutputType").Single()).InvariantEquals("Library") ? ".dll" : ".exe";

            var projectReferences = from prjRef in xdoc.Descendants(XmlHelpers.NsMsBuild + "ProjectReference").Descendants(XmlHelpers.NsMsBuild + "Project")
                                    select Guid.ParseExact(prjRef.Value, "B");
            var fbProjectReferences = from import in xdoc.Descendants(XmlHelpers.NsMsBuild + "Import")
                                      let importProject = (string) import.Attribute("Project")
                                      where importProject.InvariantStartsWith(WellKnownFolders.MsBuildProjectDir)
                                      let importProjectName = Path.GetFileNameWithoutExtension(importProject)
                                      select Guid.Parse(importProjectName);
            var allProjectReferences = projectReferences.Concat(fbProjectReferences).ToImmutableList();

            // extract binary references - both nuget and direct reference to assemblies (broken project reference)
            var binaries = from binRef in xdoc.Descendants(XmlHelpers.NsMsBuild + "Reference")
                           let assName = new AssemblyName((string) binRef.Attribute("Include")).Name
                           let maybeHintPath = binRef.Descendants(XmlHelpers.NsMsBuild + "HintPath").SingleOrDefault()
                           select new Binary(assName, null != maybeHintPath ? maybeHintPath.Value.ToUnixSeparator() : null);

            // report spurious binaries reference (System* are mostly OK)
            binaries.Where(x => x.HintPath == null && !x.AssemblyName.InvariantStartsWith("System"))
                    .ForEach(x => _logger.Warn("Spurious assembly reference {0} in project {1}", x.AssemblyName, projectFileName));

            var binaryReferences = binaries.Select(x => x.AssemblyName).ToImmutableList();

            var nugetPackages = GetNugetPackages(projectFile.Directory);
            var fbPackages = from import in xdoc.Descendants(XmlHelpers.NsMsBuild + "Import")
                             let importProject = (string) import.Attribute("Project")
                             where importProject.InvariantStartsWith(WellKnownFolders.MsBuildPackagesDir)
                             let importProjectName = Path.GetFileNameWithoutExtension(importProject)
                             select new Model.Package(importProjectName, "0.0.0");
            var packages = nugetPackages.Concat(fbPackages);
            var packageNames = packages.Select(x => x.Name).ToImmutableList();

            var project = new Project(projectGuid, projectFileName, assemblyName, extension, fxTarget, allProjectReferences, binaryReferences, packageNames);
            anthology = anthology.AddOrUpdateProject(project);

            anthology = binaries.Aggregate(anthology, (a, b) => a.AddOrUpdateBinary(b));
            anthology = packages.Aggregate(anthology, (a, p) => a.AddOrUpdatePackages(p));
            return anthology;
        }
    }
}

File: src/FsFullBuild/Workspace.fs
// Copyright (c) 2014-2015, Pierre Chalamet
// All rights reserved.
// 
// Redistribution and use in source and binary forms, with or without
// modification, are permitted provided that the following conditions are met:
//     * Redistributions of source code must retain the above copyright
//       notice, this list of conditions and the following disclaimer.
//     * Redistributions in binary form must reproduce the above copyright
//       notice, this list of conditions and the following disclaimer in the
//       documentation and/or other materials provided with the distribution.
//     * Neither the name of Pierre Chalamet nor the
//       names of its contributors may be used to endorse or promote products
//       derived from this software without specific prior written permission.
// 
// THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS" AND
// ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED
// WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
// DISCLAIMED. IN NO EVENT SHALL PIERRE CHALAMET BE LIABLE FOR ANY
// DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES
// (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
// LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND
// ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
// (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
// SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
module Workspace

open System.IO
open IoHelpers
open Env
open Configuration
open Vcs
open Anthology
open MsBuildHelpers
open System.Linq
open System.Xml.Linq
open StringHelpers
open Collections
open System

let private FindKnownProjects (repoDir : DirectoryInfo) =
    [AddExt "*" CsProj
     AddExt "*" VbProj
     AddExt "*" FsProj] |> Seq.map (fun x -> repoDir.EnumerateFiles (x, SearchOption.AllDirectories)) 
                        |> Seq.concat

let private ParseRepositoryProjects (parser) (repoRef : RepositoryId) (repoDir : DirectoryInfo) =
    repoDir |> FindKnownProjects 
            |> Seq.map (parser repoDir repoRef)

let private ParseWorkspaceProjects (parser) (wsDir : DirectoryInfo) (repos : Repository seq) = 
    repos |> Seq.map (fun x -> GetSubDirectory x.Name.Value wsDir) 
          |> Seq.filter (fun x -> x.Exists) 
          |> Seq.map (fun x -> ParseRepositoryProjects parser (RepositoryId.Bind(x.Name)) x)
          |> Seq.concat

let Init(path : string) = 
    let wsDir = DirectoryInfo(path)
    wsDir.Create()
    if IsWorkspaceFolder wsDir then failwith "Workspace already exists"
    VcsCloneRepo wsDir GlobalConfig.Repository

let Create(path : string) = 
    let wsDir = DirectoryInfo(path)
    wsDir.Create()
    if IsWorkspaceFolder wsDir then failwith "Workspace already exists"
    VcsCloneRepo wsDir GlobalConfig.Repository
    let antho = { Applications = Set.empty
                  Bookmarks = Set.empty
                  Repositories = Set.empty
                  Projects = Set.empty }
    let confDir = wsDir |> GetSubDirectory ".full-build"
    let anthoFile = confDir |> GetFile "anthology.json"
    Configuration.SaveAnthologyToFile anthoFile antho

    // FIXME
    //  create git repo in .full-build
    //       generate .gitignore
    //       add content of .full-build to git repo
    //       commit

let Index () = 
    let wsDir = WorkspaceFolder()
    let antho = LoadAnthology()
    let projects = ParseWorkspaceProjects ProjectParsing.ParseProject wsDir antho.Repositories

    // FIXME: before merging, it would be better to tell about conflicts

    // merge packages
    let foundPackages = projects |> Seq.map (fun x -> x.Packages) 
                                 |> Seq.concat
    let existingPackages = PaketParsing.ParsePaketDependencies ()
    let packagesToAdd = foundPackages |> Seq.filter (fun x -> not <| Set.contains x.Id existingPackages)
                                      |> Seq.distinctBy (fun x -> x.Id)
                                      |> Set
    PaketParsing.AppendDependencies packagesToAdd

    // merge projects
    let foundProjects = projects |> Seq.map (fun x -> x.Project)
    let newProjects = antho.Projects |> Seq.append foundProjects 
                                     |> Seq.distinctBy (fun x -> x.ProjectGuid)
                                     |> set

    let newAntho = { antho 
                     with Projects = newProjects }
    SaveAnthology newAntho

    let config = Configuration.GlobalConfig
    PaketParsing.UpdateSources config.NuGets

    Package.Simplify ()


let StringifyOutputType (outputType : OutputType) =
    match outputType with
    | OutputType.Exe -> ".exe"
    | OutputType.Dll -> ".dll"


let GenerateProjectTarget (project : Project) =
    let projectProperty = ProjectPropertyName project
    let srcCondition = sprintf "'$(%s)' != ''" projectProperty
    let binCondition = sprintf "'$(%s)' == ''" projectProperty
    let projectFile = sprintf "%s/%s/%s" MSBUILD_SOLUTION_DIR (project.Repository.Value) project.RelativeProjectFile.Value
    let binFile = sprintf "%s/%s/%s%s" MSBUILD_SOLUTION_DIR MSBUILD_BIN_OUTPUT (project.Output.Value) <| StringifyOutputType project.OutputType

    // This is the import targets that will be Import'ed inside a proj file.
    // First we include full-build view configuration (this is done to avoid adding an extra import inside proj)
    // Then we end up either importing output assembly or project depending on view configuration
    XDocument (
        XElement(NsMsBuild + "Project", 
            XElement (NsMsBuild + "Import",
                XAttribute (NsNone + "Project", "$(SolutionDir)/.full-build/views/$(SolutionName).targets"),
                XAttribute (NsNone + "Condition", "'$(FullBuild_Config)' == ''")),
            XElement (NsMsBuild + "ItemGroup",
                XElement(NsMsBuild + "ProjectReference",
                    XAttribute (NsNone + "Include", projectFile),
                    XAttribute (NsNone + "Condition", srcCondition),
                    XElement (NsMsBuild + "Project", StringifyGuid project.ProjectGuid.Value),
                    XElement (NsMsBuild + "Name", project.Output.Value)),
                XElement (NsMsBuild + "Reference",
                    XAttribute (NsNone + "Include", project.Output.Value),
                    XAttribute (NsNone + "Condition", binCondition),
                    XElement (NsMsBuild + "HintPath", binFile),
                    XElement (NsMsBuild + "Private", "true")))))

let GenerateProjects (projects : Project seq) (xdocSaver : FileInfo -> XDocument -> Unit) =
    let prjDir = WorkspaceProjectFolder ()
    for project in projects do
        let content = GenerateProjectTarget project
        let projectFile = prjDir |> GetFile (AddExt (project.ProjectGuid.Value.ToString("D")) Targets)
        xdocSaver projectFile content

let ConvertProject (xproj : XDocument) (project : Project) =
    let filterProject (xel : XElement) =
        let attr = !> (xel.Attribute (NsNone + "Project")) : string
        attr.StartsWith(MSBUILD_PROJECT_FOLDER, StringComparison.CurrentCultureIgnoreCase)

    let filterPackage (xel : XElement) =
        let attr = !> (xel.Attribute (NsNone + "Project")) : string
        attr.StartsWith(MSBUILD_PACKAGE_FOLDER, StringComparison.CurrentCultureIgnoreCase)

    let filterNuget (xel : XElement) =
        let attr = !> (xel.Attribute (NsNone + "Project")) : string
        attr.StartsWith("$(SolutionDir)\.nuget\NuGet.targets", StringComparison.CurrentCultureIgnoreCase)

    let filterNugetTarget (xel : XElement) =
        let attr = !> (xel.Attribute (NsNone + "Name")) : string
        String.Equals(attr, "EnsureNuGetPackageBuildImports", StringComparison.CurrentCultureIgnoreCase)

    let filterAssemblies (assFiles) (xel : XElement) =
        let inc = !> xel.Attribute(XNamespace.None + "Include") : string
        let assName = inc.Split([| ',' |], StringSplitOptions.RemoveEmptyEntries).[0]
        let assRef = AssemblyId.Bind (System.Reflection.AssemblyName(assName))
        let res = Set.contains assRef assFiles
        not res

    let hasNoChild (xel : XElement) =
        not <| xel.DescendantNodes().Any()

    let setOutputPath (xel : XElement) =
        xel.Value <- MSBUILD_BIN_FOLDER

    let stringifyGuid (guid : System.Guid) =
        guid.ToString("D")

    // cleanup everything that will be modified
    let cproj = XDocument (xproj)

    // remove project references
    cproj.Descendants(NsMsBuild + "ProjectReference").Remove()
    
    // remove unknown assembly references
    cproj.Descendants(NsMsBuild + "Reference").Where(filterAssemblies project.AssemblyReferences).Remove()

    // remove full-build imports
    cproj.Descendants(NsMsBuild + "Import").Where(filterProject).Remove()
    cproj.Descendants(NsMsBuild + "Import").Where(filterPackage).Remove()

    // remove nuget stuff
    cproj.Descendants(NsMsBuild + "Import").Where(filterNuget).Remove()
    cproj.Descendants(NsMsBuild + "Target").Where(filterNugetTarget).Remove()

    // set OutputPath
    cproj.Descendants(NsMsBuild + "OutputPath") |> Seq.iter setOutputPath

    // cleanup project
    cproj.Descendants(NsMsBuild + "BaseIntermediateOutputPath").Remove()
    cproj.Descendants(NsMsBuild + "ItemGroup").Where(hasNoChild).Remove()

    // add project refereces
    let afterItemGroup = cproj.Descendants(NsMsBuild + "ItemGroup").First()
    for projectReference in project.ProjectReferences do
        let prjRef = projectReference.Value.ToString("D")
        let importFile = sprintf "%s%s.targets" MSBUILD_PROJECT_FOLDER prjRef
        let import = XElement (NsMsBuild + "Import",
                        XAttribute (NsNone + "Project", importFile))
        afterItemGroup.AddAfterSelf (import)

    // add nuget references
    for packageReference in project.PackageReferences do
        let pkgId = packageReference.Value
        let importFile = sprintf "%s%s/package.targets" MSBUILD_PACKAGE_FOLDER pkgId
        let pkgProperty = PackagePropertyName pkgId
        let condition = sprintf "'$(%s)' == ''" pkgProperty
        let import = XElement (NsMsBuild + "Import",
                        XAttribute (NsNone + "Project", importFile),
                        XAttribute(NsNone + "Condition", condition))
        afterItemGroup.AddAfterSelf (import)
    cproj

let ConvertProjectContent (xproj : XDocument) (project : Project) =
    let convxproj = ConvertProject xproj project
    convxproj

let ConvertProjects (antho : Anthology) xdocLoader xdocSaver =
    let wsDir = WorkspaceFolder ()
    for project in antho.Projects do
        let repoDir = wsDir |> GetSubDirectory (project.Repository.Value)
        let projFile = repoDir |> GetFile project.RelativeProjectFile.Value 
        let xproj = xdocLoader projFile
        let convxproj = ConvertProjectContent xproj project

        xdocSaver projFile convxproj

let RemoveUselessStuff (antho : Anthology) =
    let wsDir = WorkspaceFolder ()
    for repo in antho.Repositories do
        let repoDir = wsDir |> GetSubDirectory (repo.Name.Value)
        repoDir.EnumerateFiles("*.sln", SearchOption.AllDirectories) |> Seq.iter (fun x -> x.Delete())
        repoDir.EnumerateFiles("packages.config", SearchOption.AllDirectories) |> Seq.iter (fun x -> x.Delete())
        repoDir.EnumerateDirectories("packages", SearchOption.AllDirectories) |> Seq.iter (fun x -> x.Delete(true))

let XDocumentLoader (fileName : FileInfo) =
    XDocument.Load fileName.FullName

let XDocumentSaver (fileName : FileInfo) (xdoc : XDocument) =
    xdoc.Save (fileName.FullName)

let Convert () = 
    // generate paket.dependencies and install packages
    Package.Install ()

    // generate project targets
    let antho = LoadAnthology ()
    GenerateProjects antho.Projects XDocumentSaver
    ConvertProjects antho XDocumentLoader XDocumentSaver
    RemoveUselessStuff antho

